source/common/stats/symbol_table_impl.cc

```diff
@@ -309,30 +322,68 @@ void StatNameStorage::free(SymbolTable& table) {
   bytes_.reset();
 }
 
-StatNameJoiner::StatNameJoiner(StatName a, StatName b) {
-  const uint64_t a_size = a.dataSize();
-  const uint64_t b_size = b.dataSize();
-  uint8_t* const p = alloc(a_size + b_size);
-  memcpy(p, a.data(), a_size);
-  memcpy(p + a_size, b.data(), b_size);
-}
-
-StatNameJoiner::StatNameJoiner(const std::vector<StatName>& stat_names) {
+SymbolTable::StoragePtr SymbolTableImpl::join(const std::vector<StatName>& stat_names) const {
   uint64_t num_bytes = 0;
   for (StatName stat_name : stat_names) {
     num_bytes += stat_name.dataSize();
   }
-  uint8_t* p = alloc(num_bytes);
+  auto bytes = std::make_unique<uint8_t[]>(num_bytes + StatNameSizeEncodingBytes);
+  uint8_t* p = saveLengthToBytesReturningNext(num_bytes, bytes.get());
   for (StatName stat_name : stat_names) {
     num_bytes = stat_name.dataSize();
     memcpy(p, stat_name.data(), num_bytes);
     p += num_bytes;
   }
+  return bytes;
+}
+
+StatNameList::~StatNameList() { ASSERT(!populated()); }
+
+void StatNameList::populate(const std::vector<absl::string_view>& names,
+                            SymbolTable& symbol_table) {
+  RELEASE_ASSERT(names.size() < 256, "Maximum number elements in a StatNameList exceeded");
+
+  // First encode all the names.
+  size_t total_size_bytes = 1; /* one byte for holding the number of names */
+  std::vector<SymbolEncoding> encodings;
+  encodings.resize(names.size());
+  int index = 0;
+  for (auto& name : names) {
+    SymbolEncoding encoding = symbol_table.encode(name);
+    total_size_bytes += encoding.bytesRequired();
+    encodings[index++].swap(encoding);
+  }
+
+  // Now allocate the exact number of bytes required and move the encodings
+  // into storage.
+  storage_ = std::make_unique<uint8_t[]>(total_size_bytes);
+  uint8_t* p = &storage_[0];
+  *p++ = encodings.size();
+  for (auto& encoding : encodings) {
+    p += encoding.moveToStorage(p);
+  }
+  ASSERT(p == &storage_[0] + total_size_bytes);
+}
+
+void StatNameList::foreach (const std::function<bool(StatName)>& f) const {
```

Comment:
Nit: in some places in Envoy today we use `iterate`, here we use `foreach`. Ideally we could make this an iterator and use something like `std::for_each`, but that has no short circuit I can spot. Thoughts?

Reply:
Changed to 'iterate'.